torch/csrc/jit/pickler.cpp

```diff
@@ -0,0 +1,391 @@
+    case OpCode::MARK: {
+      // Mark location of the container ivalue in the stack
+      marks_.push_back(stack_.size());
+      break;
+    }
+    case OpCode::BININT: {
+      int32_t value = read<uint32_t>();
+      stack_.emplace_back(int64_t(value));
+      break;
+    }
+    case OpCode::BINUNICODE: {
+      int32_t length = read<uint32_t>();
+      const char* characters = reinterpret_cast<const char*>(bytes_);
+      bytes_ += length;
+      stack_.emplace_back(std::string(characters, /*n=*/length));
+      break;
+    }
+    case OpCode::BINFLOAT:
+      stack_.emplace_back(readFloat());
+      break;
+    case OpCode::TUPLE: {
+      size_t start = marks_.back();
+      marks_.pop_back();
+      IValue tup = c10::ivalue::Tuple::create(
+          std::vector<IValue>(stack_.begin() + start, stack_.end()));
+      stack_.resize(start);
+      stack_.push_back(tup);
+    } break;
+    case OpCode::EMPTY_DICT:
+      stack_.emplace_back(c10::ivalue::UnorderedMap());
+      break;
+    case OpCode::APPENDS: {
+      size_t start = marks_.back();
+      marks_.pop_back();
+      auto list = stack_[start - 1].toGenericList();
+      list->elements().insert(
+          list->elements().end(), stack_.begin() + start, stack_.end());
+      stack_.resize(start);
+    } break;
+    case OpCode::SETITEMS: {
+      size_t start = marks_.back();
+      marks_.pop_back();
+      auto dict = stack_[start - 1].toGenericDict();
+      for (size_t i = start; i < stack_.size(); i += 2) {
+        dict->elements()[stack_[i]] = stack_[i + 1];
+      }
+      stack_.resize(start);
+    } break;
+    case OpCode::BINGET: {
+      stack_.push_back(memo_.at(read<uint8_t>()));
+    } break;
+    case OpCode::STOP:
+      return op;
+    case OpCode::GLOBAL: {
+      AT_ASSERT(readString() == "__main__");
+      // Push class name to stack
+      stack_.emplace_back(readString());
+    } break;
+    case OpCode::NEWOBJ: {
+      // Do nothing, should be followed by BUILD
+    } break;
+    case OpCode::BUILD: {
+      // TODO: this only works for Tensors
+      auto tensor_id = stack_.back().toInt();
+      stack_.pop_back();
+      // pop empty tuple
+      stack_.pop_back();
+      auto class_name = stack_.back().toStringRef();
+      stack_.pop_back();
+      AT_ASSERT(class_name == "TensorID");
+      stack_.emplace_back(tensor_table_.at(tensor_id));
+
+    } break;
+    default:
+      AT_ERROR("Unknown opcode for unpickling");
+  }
+
+  return op;
+}
+
+std::string Unpickler::readString(char terminator) {
+  const char* chars = reinterpret_cast<const char*>(bytes_);
+  size_t n = 0;
+  while (true) {
+    char c = chars[n];
+    if (c == '\n') {
+      break;
+    }
+    AT_ASSERT(isValidChar(c));
+    // Increment after to exclude newline from string
+    ++n;
+  }
+
+  // Increment by string length + newline char
+  bytes_ += n + 1;
+  return std::string(chars, n);
+}
+
+bool Unpickler::isValidChar(char value) {
+  return value >= 'A' && value <= 'z';
```

Comment:
GLOBAL is going to be used look up user defined classes in the future. Class_3_what is completely valid as a name but would fail this check.